packages/adapters-library/src/adapters/stargate/products/farm-v2/stargateFarmV2Adapter.ts

```diff
@@ -0,0 +1,374 @@
+    includeInUnwrap: false,
+  }
+
+  private provider: CustomJsonRpcProvider
+
+  adaptersController: AdaptersController
+
+  constructor({
+    provider,
+    chainId,
+    protocolId,
+    adaptersController,
+    helpers,
+  }: ProtocolAdapterParams) {
+    this.provider = provider
+    this.chainId = chainId
+    this.protocolId = protocolId
+    this.adaptersController = adaptersController
+    this.helpers = helpers
+  }
+
+  getProtocolDetails(): ProtocolDetails {
+    return {
+      protocolId: this.protocolId,
+      name: 'Stargate Farm V2',
+      description:
+        'Stargate is a fully composable liquidity transport protocol that lives at the heart of Omnichain DeFi',
+      siteUrl: 'https://stargate.finance/',
+      iconUrl: 'https://stargate.finance/favicons/favicon-light.svg',
+      positionType: PositionType.Staked,
+      chainId: this.chainId,
+      productId: this.productId,
+    }
+  }
+
+  @CacheToFile({ fileKey: 'pool' })
+  async getProtocolTokens(): Promise<ProtocolToken<AdditionalMetadata>[]> {
+    const { stargateStakingAddress } = staticChainDataV2[this.chainId]!
+    const stakingContract = StargateStaking__factory.connect(
+      stargateStakingAddress,
+      this.provider,
+    )
+
+    const poolAddresses = await stakingContract['tokens()']()
+
+    return await Promise.all(
+      poolAddresses.map(async (poolAddress) => {
+        const protocolTokenPromise = this.helpers.getTokenMetadata(
+          getAddress(poolAddress),
+        )
+
+        const rewarderAddress = await stakingContract.rewarder(poolAddress)
+
+        const multiRewarderContract = StargateMultiRewarder__factory.connect(
+          rewarderAddress,
+          this.provider,
+        )
+
+        const rewardTokenAddresses = await multiRewarderContract.rewardTokens()
+
+        return {
+          ...(await protocolTokenPromise),
+          rewarderAddress,
+          rewardTokens: await Promise.all(
+            rewardTokenAddresses.map(async (rewardTokenAddress) =>
+              this.helpers.getTokenMetadata(getAddress(rewardTokenAddress)),
+            ),
+          ),
+        }
+      }),
+    )
+  }
+
+  private async getProtocolTokenByAddress(
+    protocolTokenAddress: string,
+  ): Promise<ProtocolToken<AdditionalMetadata>> {
+    return this.helpers.getProtocolTokenByAddress({
+      protocolTokens: await this.getProtocolTokens(),
+      protocolTokenAddress,
+    })
+  }
+
+  async getPositions({
+    userAddress,
+    blockNumber,
+    protocolTokenAddresses,
+  }: GetPositionsInput): Promise<ProtocolPosition[]> {
+    const { stargateStakingAddress } = staticChainDataV2[this.chainId]!
+    const protocolTokens = await this.getProtocolTokens()
+
+    return await filterMapAsync(protocolTokens, async (protocolToken) => {
+      if (
+        protocolTokenAddresses &&
+        !protocolTokenAddresses.includes(protocolToken.address)
+      ) {
+        return undefined
+      }
```

Comment:
Do you see a future where we can drop this filter? Wondering if the filtering could be done outside....

Reply:
To be honest, I can't.

If we do it outside that means the return function will be something like `Promise<ReturnType | undefined>`, which doesn't make a lot of sense.

Moreover, we use that pattern pretty often internally as well, which means that if we don't get rid of it quickly and we chain the result to another array method, the first thing we'll have to do is check if it's undefined.

It's unfortunate that JS doesn't come with an array method that does that automatically.